Add new application exit reason RESOURCE_CONSTRAINT

mvorobio opened the pull request:
Fixes #[0273](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0273-webengine-projection-mode.md)

This PR is **not ready** for review.

### Risk
This PR makes **minor** API changes.
Rpc_spec https://github.com/mvorobio/rpc_spec/pull/1

### Testing Plan
Use scripts from https://github.com/mvorobio/sdl_atf_test_scripts/pull/1

### Summary
The PR adds RESOURCE_CONSTRAINT to HMI API and its processing to onExitApplicationNotification, updates unit tests.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

ShobhitAd committed:
Fix/SDL does not send BC.PolicyUpdate to HMI if no apps are registered (#3280)

* Temporary fix for allowing core to send the policyupdate rpc

* Remove CanUpdate call in PolicyManagerImpl::StartPTExchange and restore CanUpdate behavior

* Fix regular policy unit tests

* Fix extern policy unit tests

ShobhitAd committed:
Fix/Perform hash check before copying over mobile api files (#3288)

* Compare MOBILE_API files in the submodule directory against the ones in the interfaces directory before copying

* Use cmake copy_if_different command instead of cmp

* Remove trailing whitespace

ShobhitAd committed:
fix/Move AppProperties struct to the Common interface (#3286)

* Move AppProperties struct to the Common interface

* Remove unnecessary response params from SetAppProperties and GetAppProperties response

* Add missing description tag

* Fix description tag in last commit

iCollin committed:
Fix OnSystemRequest(LOCK_SCREEN_ICON_URL) sending (#3271) (#3289)

There was noticed an issue that sometimes Policy Manager
component is returning an empty url for lock screen icon.
The root cause of that is the data races between thread
which is sending OnSystemRequest notification and backupper
thread which overwrites data in the database. As these two
threads may have an intersection, there could be a possible
situation that backuper thread erased data from `endpoints`
table and another thread is using SQL query to select the
data from the same table. As a result, second thread receives
an empty result as the data was erased and the new one was not
inserted yet by backupper thread. After that, SDL just does
not send OnSystemRequest as the URL is empty.
To fix that issue, policy component has been updated to use
internal policy cache instead of direct access to policy
database each time to get URL. This will guarantee that
a sending thread will get an up-to-date infromation
required for OnSystemRequest notification.
Also, code cleanup has been done in order to remove all
functions which become unused after this fix.

Co-authored-by: Andrii Kalinich (GitHub) <AKalinich@luxoft.com>

SNiukalov committed:
Fix/consider timeouts during streaming switching (#3293)

* Consider video and audio status before send end of service

Co-authored-by: sniukalov <sniukaov@luxoft.com>

ShobhitAd committed:
Fix/Checking PTU triggers when no apps are registered in regular policy mode (#3294)

* Trigger PTExchange when Device consent is enabled

* Move StartPTExchange call to BC.OnReady

* StartPTExchange is a protected function. Add TriggerPTUIfNeeded to the policy_manager as well

* Fixed GetPolicyHandler call in BC.OnReady

* Add mock definitions

* Fix failed ut

* Handle PolicyUpdate error response

* Add reset retry sequence

* Revert "Move StartPTExchange call to BC.OnReady"

This reverts commit 7c53978b2e7a42ed51cb856d054dce88e040e8ea.

Co-authored-by: JackLivio <jack@livio.io>

YarikMamykin committed:
Fix sdl crash when policies disabled (#3292)

* Fix usage non-valid policy_manager pointer in case of EnablePolicy=false

* Add unit tests for modified methods

Co-authored-by: Jacob Keeler <jacob.keeler@livioradio.com>

iCollin committed:
Update start_external_proprietary.sh (#3299)

* Update start_external_proprietary.sh

* Update src/appMain/start_external_proprietary.sh

Co-Authored-By: Shobhit Adlakha <ShobhitAd@users.noreply.github.com>

Co-authored-by: Shobhit Adlakha <ShobhitAd@users.noreply.github.com>

iCollin committed:
maintain signature for GetLockScreenIconUrl (#3303)

iCollin committed:
move submodule to own org

ShobhitAd committed:
Fix/Cloud and Web-engine apps disappear after refreshing the page (#3308)

* Send updateapplist when hmi cooperation starts

* Remove extraneous logs

Jack-Byrne committed:
Update KMS count last update took place at (#3311)

ShobhitAd committed:
Fix/Create generic handler in sample policy manager for different system request types (#3315)

* Create a generic handler to handle different system request types

* Change function name from crypt to encrypt

SNiukalov committed:
Update path to smartDeviceLink.ini (#3310)

Co-authored-by: sniukalov <sniukaov@luxoft.com>

Commit:
Remove redundant GetSystemInfo sending (#3305)

zhouxin627 committed:
Fix for video stream failed by switched of connection via BT+WIFI (#3290)

Co-authored-by: jiazhouyang <jiazhouyang@iauto.com>
Co-authored-by: JackLivio <jack@livio.io>

Jack-Byrne committed:
Fix remove_if in ClearTimerPool (#3322)

Ypostolov committed:
Fix SDL doesn't send EndService for Video/Audio Service upon IGNITION_OFF #1409 (#2425)

* Fix SDL doesn't send EndService for Video/Audio Service upon IGNITION_OFF

zhouxin627 committed:
fix APTDisplayText covert bug (#3291)

Co-authored-by: xiatian <xiatian@iauto.com>

mvorobio committed:
Add new application exit reason RESOURCE_CONSTRAINT

Add RESOURCE_CONSTRAINT to HMI API and its processing to
OnExitApplicationNotification, update unit tests

AByzhynar commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/on_exit_application_notification.cc:
```diff
@@ -112,6 +112,15 @@ void OnExitApplicationNotification::Run() {
+              app_id, AppInterfaceUnregisteredReason::RESOURCE_CONSTRAINT);
+      SendNotificationToMobile(message);
+      application_manager_.UnregisterApplication(app_id, Result::SUCCESS);
+      application_manager_.connection_handler().CloseRevokedConnection(app_id);
```
@mvorobio What is the reason to call this function?

mvorobio replied:
@AByzhynar CloseRevokedConnection is called to close WebSocket connection associated to the web app HMI wants to exit.

AByzhynar replied:
@mvorobio I understand what the function does. I am asking why did you decide to close connection? I can't see any requirements.  So for me it looks just doing something which is not required and requested by the customer.

mvorobio replied:
@AByzhynar Summary of https://adc.luxoft.com/jira/browse/FORDTCN-5000 implies this. Not sure if it is correct interpretation of "there is a way for the HMI to be able to unregister and **disconnect** an app in case of performance issues"

AByzhynar replied:
@mvorobio Don't interpret. Interpretations can differ. Read it as is: `unregister and disconnect an app` but you disconnect the entire transport, don't you?

SNiukalov replied:
@AByzhynar @mvorobio 
I believe, in this case we disconnect only session for this app, other sessions and listener will be continue working.

SNiukalov committed:
Fix/Send OnAppPropertiesChange in case one of nicknames is removed (#3336)

* Update compare for the nicknames

* UTs update

Co-authored-by: sniukalov <sniukaov@luxoft.com>

SNiukalov committed:
Fix/Deadlock during transport re-connect via USB (#3332)

* Rework the use of AbortConnection for a usb connection

Sometimes we get a deadlock because we use the same lock in the same thread.
The situation is possible when calling the SendData, OnOutTransfer or AbortConnection functions.
They block the mutex at the start.
If in the PostOutTransfer function, which is called in SendData or indirectly in OnOutTransfer via PopOutMessage, we get the error LIBUSB_ERROR_NO_DEVICE,
this will call AbortConnection.
Which will cause a deadlock.

* fixup! Rework the use of AbortConnection for a usb connection

Co-authored-by: sniukalov <sniukaov@luxoft.com>

YarikMamykin committed:
Hotfix/3327 sdl sends generic error if vehicle data not available (#3335)

* Fix response message structure of GetVehicleData request

In case of GetVehicleData error response from HMI,
SDL put only part of message to event. This way
event was not processed properly and GENERIC_ERROR
was sent to HMI.

This commit fixes error response message structure.

* Update existing unit tests

* fixup! Update existing unit tests

Co-authored-by: YarikMamykin <ymamykin@gmail.com>

Commit:
Do not send two END_SERVICE requests during streaming switching between apps  (#3334)

* Protect navi_app_to_stop_

* Prevent EndService double sending

Remove app_id from navi_service_status_ before unregistering the app
in case EndServiceAck did not come for at least one of the services,
in order to avoid sending the EndService again during unregistration

jacobkeeler committed:
Fix PROPRIETARY and HTTP retry sequences (#3329)

* Fix PROPRIETARY and HTTP retry sequences

For PROPRIETARY policy updates, Core is responsible for the retry sequence and cycling through update URLs. This fixes the PROPRIETARY flow with a few specific changes:

1. Apps for PTU are chosen randomly using the existing process, but after an application is chosen, all URLs that are associated with that application will be attempted before moving onto the next application.
2. When an OnSystemRequest(PROPRIETARY) is received from the HMI, Core will ignore the `appID` and `url` properties and populate them itself. This is because Core is responsible for providing these fields in PROPRIETARY mode.

* Add caching for PTU URLs if provided by HMI

* Fix bug in HTTP mode

iCollin committed:
StartEndStreamTimer in StopNaviService (#3325)

jacobkeeler committed:
Move cached app logic to ChoosePTUApplication

Also add macro guards for uses of last_ptu_app_id_

mvorobio committed:
Add web socket connection closing on RESOURCE_CONSTRAINT

AKalinich-Luxoft committed:
Fix UpdateAppList/UpdateDeviceList spam (#3339)

* Fix UpdateAppList/UpdateDeviceList spam

There was noticed a continuous spam of update app
and device list requests from SDL to HMI even if
thre is no any devices connected and apps registered.
After analysis, there was found a couple of places
where logic should be updated:
- In OnHMIStartedCooperation don't send update app
list after refreshing cloud app information as this
function able to trigger the same request itself when
needed
- When update app list timer is expired, don't send
update app list if no any apps were registered or
unregistered while timer was running
- In transport manager, don't raise OnDeviceListUpdated
event if no any devices were added/removed
- When WebEngine device is created, it makes sense to
connect it automatically and set its status to CONNECTED
as its internal communication between SDL and HMI
- In transport adapter, don't notify listeners via
OnConnectionStatusUpdated() callbacks if device connection
status was not actually changed

All these actions together prevents all possible reasons
of UpdateAppList/UpdateDeviceList spam observed before fix.

* fixup! Fix UpdateAppList/UpdateDeviceList spam

* fixup! fixup! Fix UpdateAppList/UpdateDeviceList spam

* fixup! fixup! fixup! Fix UpdateAppList/UpdateDeviceList spam

jacobkeeler committed:
Merge pull request #3341 from smartdevicelink/fix/external_policy_build

Move cached app logic to ChoosePTUApplication

mvorobio committed:
Use fork of rpc_spec

mvorobio commented:
The changes moved to https://github.com/LuxoftSDL/sdl_core/pull/18
